src/tangents.jl

```diff
@@ -997,30 +1025,58 @@ function _diff_internal(c::MaybeCache, p::P, q::P) where {P<:SimpleVector}
     return t
 end
 function _diff_internal(c::MaybeCache, p::P, q::P) where {P<:Union{Tuple,NamedTuple}}
-    tangent_type(P) == NoTangent && return NoTangent()
-    return _map((p, q) -> _diff_internal(c, p, q), p, q)
+    TP = tangent_type(P)
+    TP == NoTangent && return NoTangent()
+    return _map((p, q) -> _diff_internal(c, p, q), p, q)::TP
 end
 
 function _containerlike_diff(c::MaybeCache, p::P, q::P) where {P}
-    if ismutabletype(P)
-        t = tangent_type(P)()
-        c[(p, q)] = t
-    end
-    diffed_fields = map(fieldnames(P)) do f
-        if isdefined(p, f) && isdefined(q, f)
-            return _diff_internal(c, getfield(p, f), getfield(q, f))
-        elseif !isdefined(p, f) && !isdefined(q, f)
-            return FieldUndefined()
+    return _containerlike_diff_cartesian(c, p, q, Val(ismutabletype(P)), Val(fieldcount(P)))
+end
+@generated function _containerlike_diff_cartesian(
+    c::MaybeCache, p::P, q::P, ::Val{mutable}, ::Val{nfield}
+) where {P,mutable,nfield}
+    quote
+        t = if mutable
+            _t = tangent_type(P)()
+            c[(p, q)] = _t
+            _t
         else
-            throw(error("Unhandleable undefinedness"))
+            nothing
         end
+        Base.Cartesian.@nif(
+            $(nfield + 1),
+            n -> let
+                defined_p = isdefined(p, n)
```

Comment:
This new implementation implicitly assumes the ordering of fields is the same, whereas the current code doesn't. So, this might not be correct in all cases.

Reply:
I overlooked that. We should insert a check to prevent future changes from accidentally breaking the assumption.